### PR TITLE
add state and arrow requests

### PR DESCRIPTION
**Pull request by arfio**
This pull request add the implementation for the arrows and states routes.

This change follows the following TSP specification:
- https://eclipse-cdt-cloud.github.io/trace-server-protocol/#tag/TimeGraph/operation/getStates
- https://eclipse-cdt-cloud.github.io/trace-server-protocol/#tag/TimeGraph/operation/getArrows

### CODE REVIEW

**Review by MatthewKhouzam — CHANGES_REQUESTED** [review]
Hi Arnaud, this looks great, but can we have a sanity test for it? thanks!

File: tsp/tsp_client.py [hunk 1]
```diff
@@ -345,23 +348,83 @@ def fetch_timegraph_tree(self, exp_uuid, output_id, parameters=None):
...
+        params = parameters
+        if parameters is None:
+            params = {
+                "parameters": { }
+            }
+
+        response = requests.post(api_url, json=params, headers=headers)
+
+        if response.status_code == 200:
+            response_size = len(bytes(response.text, 'utf-8'))
```
Comment (bhufmann) [inline tsp/tsp_client.py:391]: response_size can be removed because not used.

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the contribution. It looks good. I have a small comment (unused variable) though.

@arfio Do you plan to provide a follow-up PR on top of this PR to add a command-line interface for these added tsp_client methods?

**Comment by bhufmann** [comment]
@arfio any news for this PR?

File: tsp/tsp_client.py (hunk 1 above)
Comment (arfio) [reply tsp/tsp_client.py:391]: I removed all the references to response_size.

**Comment by arfio** [comment]
I did not plan to add a follow-up PR to add additional CLI commands for these. I do not know if it is really needed as this is not really human readable data.

**Comment by bhufmann** [comment]
> I did not plan to add a follow-up PR to add additional CLI commands for these. I do not know if it is really needed as this is not really human readable data.

Thanks for the answer. I agree it's harder to make us of it from the CLI. I'm ok to merge this PR without it and a CLI can be added later if ever there is a need for it. Please update this PR for the small comment I provided. Thanks.

**Comment by arfio** [comment]
I believe I have already made the change needed for your comment ?

**Review by bhufmann — CHANGES_REQUESTED** [review]
I found some issues with the tests that did get unnoticed when merge an earlier PR. I have fix for that with PR #81. After merging that PR you will need to rebase and also update your new tests to apply the same. Also, I found 2 issues in the newly fetch method for states and arrows. Please have a look.

**Comment by bhufmann** [comment]
@arfio PR #81 is merged and you can rebase this PR.

**Comment by kavehshahedi** [comment]
@arfio
Hey Arnaud,

Did you have a chance to complete this PR? I tested it, and it seems to be working fine. Could you please update it so it can be merged into the master?

Thanks!

File: test_tsp.py [hunk 2]
```diff
@@ -501,6 +502,87 @@ def test_fetch_virtual_table_lines(self, ust):
...
+
+        response = self.tsp_client.fetch_experiment_outputs(experiment_uuid)
+        output_id = TIMEGRAPH_DP_ID
+        status = ResponseStatus.RUNNING.name
+        while status == ResponseStatus.RUNNING.name:
+            time.sleep(1)
+            response = self.tsp_client.fetch_timegraph_tree(
+                experiment_uuid, output_id)
+            assert response.model is not None
+            status = response.model.status.upper()
```
[the remark this answers is not in the document]
Comment (arfio) [reply test_tsp.py]: Done.

File: test_tsp.py [hunk 3]
```diff
@@ -501,6 +502,87 @@ def test_fetch_virtual_table_lines(self, ust):
...
```
[the remark this answers is not in the document]
Comment (arfio) [reply test_tsp.py]: Done

File: test_tsp.py [hunk 4]
```diff
@@ -501,6 +502,87 @@ def test_fetch_virtual_table_lines(self, ust):
...
+        response = self.tsp_client.open_trace(os.path.basename(kernel), kernel)
+        traces.append(response.model.UUID)
+        response = self.tsp_client.open_experiment(
+            os.path.basename(kernel), traces)
+        assert response.status_code == 200
+        experiment_uuid = response.model.UUID
+
+        response = self.tsp_client.fetch_experiment_outputs(experiment_uuid)
+        output_id = TIMEGRAPH_DP_ID
+        status = ResponseStatus.RUNNING.name
```
[the remark this answers is not in the document]
Comment (arfio) [reply test_tsp.py]: Done

File: test_tsp.py [hunk 5]
```diff
@@ -501,6 +502,87 @@ def test_fetch_virtual_table_lines(self, ust):
...
```
[the remark this answers is not in the document]
Comment (arfio) [reply test_tsp.py]: Done

File: test_tsp.py [hunk 6]
```diff
@@ -501,6 +502,87 @@ def test_fetch_virtual_table_lines(self, ust):
...
+        while status == ResponseStatus.RUNNING.name:
```
[the remark this answers is not in the document]
Comment (arfio) [reply test_tsp.py]: Done

File: tsp/tsp_client.py [hunk 7]
```diff
@@ -363,6 +367,62 @@ def fetch_timegraph_tree(self, exp_uuid, output_id, parameters=None):
...
+            params = {
+                "parameters": { }
+            }
+
+        response = requests.post(api_url, json=params, headers=headers)
+
+        if response.status_code == 200:
+            return TspClientResponse(GenericResponse(json.loads(response.content.decode('utf-8')),
+                                                     ModelType.TIME_GRAPH_STATE),
+                                     response.status_code)
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client.py]: Done

File: tsp/tsp_client.py [hunk 8]
```diff
@@ -363,6 +367,62 @@ def fetch_timegraph_tree(self, exp_uuid, output_id, parameters=None):
...
+        params = parameters
+        if parameters is None:
+            params = {
+                "parameters": { }
+            }
+
+        response = requests.post(api_url, json=params, headers=headers)
+
+        if response.status_code == 200:
+            return TspClientResponse(GenericResponse(json.loads(response.content.decode('utf-8')),
```
[the remark this answers is not in the document]
Comment (arfio) [reply tsp/tsp_client.py:419]: Done

**Comment by arfio** [comment]
Hey Kaveh,

Thanks for reminding me, I had corrected things earlier but forgot to push the changes.

It should be ok now !

**Review by MatthewKhouzam — APPROVED** [review]
This should help TMLL and many other things!

**Pull request merged**